File: apollo-android-support/src/main/java/com/apollographql/apollo/ApolloCallback.java
package com.apollographql.apollo;

import android.os.Handler;
import android.os.Looper;

import com.apollographql.apollo.api.Response;
import com.apollographql.apollo.exception.ApolloException;
import com.apollographql.apollo.exception.ApolloHttpException;
import com.apollographql.apollo.exception.ApolloNetworkException;
import com.apollographql.apollo.exception.ApolloParseException;

import org.jetbrains.annotations.NotNull;

import static com.apollographql.apollo.api.internal.Utils.checkNotNull;

/**
 * <p>Android wrapper for {@link ApolloCall.Callback} to be operated on specified {@link Handler}</p>
 *
 * <b>NOTE:</b> {@link #onHttpError(ApolloHttpException)} will be called on the background thread if provided handler is
 * attached to the main looper. This behaviour is intentional as {@link ApolloHttpException} internally has a reference
 * to raw {@link okhttp3.Response} that must be closed on the background, otherwise it throws {@link
 * android.os.NetworkOnMainThreadException} exception.
 */
public final class ApolloCallback<T> extends ApolloCall.Callback<T> {
  final ApolloCall.Callback<T> delegate;
  private final Handler handler;

  /**
   * Wraps {@code callback} to be be operated on specified {@code handler}
   *
   * @param callback original callback to delegates calls
   * @param handler  the callback will be run on the thread to which this handler is attached
   */
  public static <T> ApolloCallback<T> wrap(@NotNull ApolloCall.Callback<T> callback, @NotNull Handler handler) {
    return new ApolloCallback<>(callback, handler);
  }

  /**
   * @param callback original callback to delegates calls
   * @param handler  the callback will be run on the thread to which this handler is attached
   */
  public ApolloCallback(@NotNull ApolloCall.Callback<T> callback, @NotNull Handler handler) {
    this.delegate = checkNotNull(callback, "callback == null");
    this.handler = checkNotNull(handler, "handler == null");
  }

  @Override public void onResponse(@NotNull final Response<T> response) {
    handler.post(new Runnable() {
      @Override public void run() {
        delegate.onResponse(response);
      }
    });
  }

  @Override public void onStatusEvent(@NotNull final ApolloCall.StatusEvent event) {
    handler.post(new Runnable() {
      @Override public void run() {
        delegate.onStatusEvent(event);
      }
    });
  }

  @Override public void onFailure(@NotNull final ApolloException e) {
    handler.post(new Runnable() {
      @Override public void run() {
        delegate.onFailure(e);
      }
    });
  }

  @Override public void onHttpError(@NotNull final ApolloHttpException e) {
    if (Looper.getMainLooper() == handler.getLooper()) {
      delegate.onHttpError(e);
    } else {
      handler.post(new Runnable() {
        @Override public void run() {
          delegate.onHttpError(e);
        }
      });
    }
  }

  @Override public void onNetworkError(@NotNull final ApolloNetworkException e) {
    handler.post(new Runnable() {
      @Override public void run() {
        delegate.onNetworkError(e);
      }
    });
  }

  @Override public void onParseError(@NotNull final ApolloParseException e) {
    handler.post(new Runnable() {
      @Override public void run() {
        delegate.onParseError(e);
      }
    });
  }
}

File: apollo-android-support/src/main/kotlin/com/apollographql/apollo/ApolloCallback.kt
package com.apollographql.apollo

import android.os.Handler
import com.apollographql.apollo.api.internal.Utils.__checkNotNull
import android.os.Looper
import com.apollographql.apollo.ApolloCallback
import com.apollographql.apollo.api.Response
import com.apollographql.apollo.exception.ApolloException
import com.apollographql.apollo.exception.ApolloHttpException
import com.apollographql.apollo.exception.ApolloNetworkException
import com.apollographql.apollo.exception.ApolloParseException

/**
 *
 * Android wrapper for [ApolloCall.Callback] to be operated on specified [Handler]
 *
 * **NOTE:** [.onHttpError] will be called on the background thread if provided handler is
 * attached to the main looper. This behaviour is intentional as [ApolloHttpException] internally has a reference
 * to raw [okhttp3.Response] that must be closed on the background, otherwise it throws [ ] exception.
 */
class ApolloCallback<T>(callback: ApolloCall.Callback<T>, handler: Handler) : ApolloCall.Callback<T>() {
  val delegate: ApolloCall.Callback<T>
  private val handler: Handler
  override fun onResponse(response: Response<T>) {
    handler.post { delegate.onResponse(response) }
  }

  override fun onStatusEvent(event: ApolloCall.StatusEvent) {
    handler.post { delegate.onStatusEvent(event) }
  }

  override fun onFailure(e: ApolloException) {
    handler.post { delegate.onFailure(e) }
  }

  override fun onHttpError(e: ApolloHttpException) {
    if (Looper.getMainLooper() == handler.looper) {
      delegate.onHttpError(e)
    } else {
      handler.post { delegate.onHttpError(e) }
    }
  }

  override fun onNetworkError(e: ApolloNetworkException) {
    handler.post { delegate.onNetworkError(e) }
  }

  override fun onParseError(e: ApolloParseException) {
    handler.post { delegate.onParseError(e) }
  }

  companion object {
    /**
     * Wraps `callback` to be be operated on specified `handler`
     *
     * @param callback original callback to delegates calls
     * @param handler  the callback will be run on the thread to which this handler is attached
     */
    fun <T> wrap(callback: ApolloCall.Callback<T>, handler: Handler): ApolloCallback<T> {
      return ApolloCallback(callback, handler)
    }
  }

  /**
   * @param callback original callback to delegates calls
   * @param handler  the callback will be run on the thread to which this handler is attached
   */
  init {
    delegate = __checkNotNull(callback, "callback == null")
    this.handler = __checkNotNull(handler, "handler == null")
  }
}